### PR TITLE
fix(amf): AMF is accepting the PDU Session when a PDU Request with DNN as empty string is been received from UE.

### LINKED ISSUE #10477: [AMF] AMF is accepting the PDU-Session when a PDU-Request with DNN as empty string is been received from UE.
Your Environment

- Version: 1.7.0-1637360209-fcdd682f
- Affected Component: Access Gateway
- Affected Subcomponent: AGW service
- Deployment Environment: Vagrant (AGW)

Describe the Issue

AMP is sending accept message for a PDU-Session request with empty DNN-Name is given.

**To Reproduce**
1. Configure subscriber without any DNN/APN name.
2. Send PDU session request from UE with DNN-Name as empty.
3. Check that PDU-Session is getting rejected are notified.
4.  PDU-Session is getting accepted and session is getting established.

Expected behavior

PDU-Session reject or notification should be sent and PDU-Session should not get established.

Screenshots

Screenshot of PDU-Session Request with DNN-IE:

PDU-Accept for Request:

Syslogs where DNN string is taken as '0'

[https://app.zenhub.com/files/170803235/a3927dc7-6fd2-4cd5-b589-77f243f79116/download](https://app.zenhub.com/files/170803235/a3927dc7-6fd2-4cd5-b589-77f243f79116/download)

### LINKED ISSUE #10479: [AMF] AMF is sending PDU-Reject with cause 27 for PDU-Release command from UE.
Prior to submitting an issue, check to see if one has already been created. If there is currently an open issue, add a thumbs-up emoji to identify that it is also affecting you.

### Your Environment

- Version: 1.7.0-1637360209-fcdd682f
- Affected Component: Access Gateway
- Affected Subcomponent: AGW service
- Deployment Environment: Vagrant (AGW)

### Describe the Issue
[AMF] AMF is sending PDU-Reject with cause 27 for PDU-Release command from UE.

**To Reproduce**
1. Add subscriber without attaching APN.
2. Send PDU session request from UE with DNN-Name as empty.
3. Check that PDU-Session is getting rejected are notified.
4. PDU-Session is getting accepted. For PDU-Release from UE AMF is sending PDU-Reject with cause 27.

**Expected behavior**
PDU-Reject should not be sent for PDU-Release command.

**Screenshots**

Subscriber Configuration:

### PR DESCRIPTION
**Pull request by sreedharkumartn**
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

## Summary
Issues handled:
1. AMF is accepting the PDU-Session when a PDU-Request with DNN as empty string is been received from UE. (Zenhub task ticket #10477).
2. AMF is sending PDU-Reject with cause 27 for PDU-Release command from UE. (Zenhub task ticket #10479).

## Test Plan
Fix for Issue 1.
Subscriber dnn(apn) info:
subscriber
 
UERANSIM dnn(apn) info:
ueransim_cause_1

Pcap Snap:
Packet_cause1_23rdnov

Log:
[mme_cause_1_resolved_23nov.log](https://github.com/magma/magma/files/7586937/mme_cause_1_resolved_23nov.log)

Fix for Issue 2.
Pcap Snap:
Packet_cause2_23rdnov

Log:
[mme_cause_2_resolved_23nov.log](https://github.com/magma/magma/files/7586973/mme_cause_2_resolved_23nov.log)

UT cases Validated:
UT cases

Checked Normal Registration accept, PDU session establishment request and PDU session establishment release with UERANSIM
Pcap Snap:
Packet_23rdnov

Log:
[mme_23nov.log](https://github.com/magma/magma/files/7587005/mme_23nov.log)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp [hunk 1]
```diff
@@ -593,7 +595,7 @@ M5GSmCause amf_smf_get_smcause(amf_ue_ngap_id_t ue_id, ULNASTransportMsg* msg) {
   the external DNN because the DNN was not included
   although required or if the DNN could not be resolved.
   */
-  if (msg->dnn.len == 0 &&
+  if (msg->dnn.len <= 1 &&
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:598]: what is the reason for this change? should not this be instead `msg->dnn.len == 0 || (ue_context->amf_context.apn_config_profile.nb_apns == 0)`

**Review by ulaskozat — CHANGES_REQUESTED** [review]
see my comment

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp (hunk 1 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:598]: MicrosoftTeams-image (5)
msg->dnn.len is the length of dnn contents. DNN value will have the first byte that tells the DNN length and then DNN value. In this case msg->dnn.len is 1 means no DNN content. Minimum value should be more than 1. 

```should not this be instead msg->dnn.len == 0 || (ue_context->amf_context.apn_config_profile.nb_apns == 0)```
If UE sends empty DNN or DNN IE not present. As per PR #9645 default DNN is picking from apn_config_profile.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:598]: what about the AND vs OR statement here. Should not it be OR? If there is no apn profile for UE, it should be rejected.

Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:598]: If there is no apn profile for UE and UE sends DNN. As per PR #9645 PDU session will be rejected with cause 91: DNN not supported or not subscribed.

Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:598]: image
@ulaskozat these are the cases we are handling and their current behavior. Please review it and provide the feedback. 
Above the case is for the Cause 27, #9645 PR handling other scenarios.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:598]: @sreedharkumartn where is cause #91 implemented?

**Review by ulaskozat — APPROVED** [review]
Approving this. But don't forget to answer the question.

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp (hunk 1 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:598]: Cause #91 was implemented in ```lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp``` in PR #9645.
image
image

**Pull request merged**